File: MagicHax/android/app/src/main/java/me/nathanp/magiccreatures/viewmodel/CreatureBuilderViewModel.java
package me.nathanp.magiccreatures.viewmodel;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import me.nathanp.magiccreatures.model.Card;
import me.nathanp.magiccreatures.model.CardRepository;
import me.nathanp.magiccreatures.model.Creature;
import me.nathanp.magiccreatures.model.Trait;
import me.nathanp.magiccreatures.model.TraitRepository;

public class CreatureBuilderViewModel extends ViewModel {

    private int startingPoints = 0;
    private Set<Trait> selectedTraits = new HashSet<>();
    private Set<Card> selectedCards = new HashSet<>();

    CardRepository cardRepository = new CardRepository();
    TraitRepository traitRepository = new TraitRepository();

    private MutableLiveData<Creature> mCreatureLiveData = new MutableLiveData<>();
    private MutableLiveData<Integer> remainingPointsLiveData = new MutableLiveData<>();

    private LiveData<ArrayList<Trait>> availableTraits;
    private LiveData<ArrayList<Card>> availableCards;

    public void setCreature(Creature creature) {
        mCreatureLiveData.setValue(creature);
    }

    public LiveData<Creature> getCreature() {
        return mCreatureLiveData;
    }

    public LiveData<List<Trait>> getTraits() {
        return traitRepository.getAllTraits();
    }

    public LiveData<List<Card>> getCards() {
        return cardRepository.getAllCards();
    }

    public void selectCard(Card c) {
        selectedCards.add(c);
        c.setSelected(true);

    }

    public void unselectCard(Card c) {
        selectedCards.remove(c);
        c.setSelected(false);
    }

    public void selectTrait(Trait t) {
        selectedTraits.add(t);
        t.setSelected(true);
    }

    public void unselectTrait(Trait t) {
        selectedTraits.remove(t);
        t.setSelected(false);
    }

    private int remainingPoints() {
        return startingPoints - totalCostOfTraits(selectedTraits) - totalCostOfCards(selectedCards);
    }

    private int totalCostOfCards(Set<Card> cards) {
        int total = 0;
        for (Card card : cards) {
            total += card.cost;
        }
        return total;
    }

    private int totalCostOfTraits(Set<Trait> traits) {
        int total = 0;
        for (Trait trait : traits) {
            total += trait.cost;
        }
        return total;
    }
}

File: MagicHax/android/app/src/main/java/me/nathanp/magiccreatures/viewmodel/CreatureBuilderViewModel.kt
package me.nathanp.magiccreatures.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import me.nathanp.magiccreatures.model.CardRepository
import me.nathanp.magiccreatures.model.Creature
import me.nathanp.magiccreatures.model.Resources.Types.getRadioIdFromType1
import me.nathanp.magiccreatures.model.Resources.Types.getRadioIdFromType2
import me.nathanp.magiccreatures.model.Resources.Types.getTypeFromRadioId
import me.nathanp.magiccreatures.model.Trait
import me.nathanp.magiccreatures.model.TraitRepository

class CreatureBuilderViewModel : ViewModel() {
    private val TAG = javaClass.simpleName
    private val startingPoints = 0
    private val cardRepository = CardRepository()
    private val traitRepository = TraitRepository()
    private val _creature = MutableLiveData<Creature>()
    private val _remainingPoints = MutableLiveData<Int>()
    // Public access to LiveData

    var creature: LiveData<Creature> = _creature
    var points: LiveData<Int> = _remainingPoints
    var cards = cardRepository.allCards
    var traitMap = traitRepository.traitMap

    var creatureImage = Transformations.map(_creature) { it.drawableId }

    fun setCreature(c: Creature?) {
        _creature.value = c
    }

    var type1: Int
        get() = getRadioIdFromType1(creature.value!!.type1)
        set(id) {
            _creature.value!!.type1 = getTypeFromRadioId(id)
            _creature.value = _creature.value
        }

    var type2: Int
        get() = getRadioIdFromType2(creature.value!!.type2)
        set(id) {
            _creature.value!!.type2 = getTypeFromRadioId(id)
            _creature.value = _creature.value
        }

    fun selectTrait(t: Trait): Boolean {
        val added = _creature.value!!.addTrait(t.name)
        if (added) _creature.value = _creature.value
        return added
    }

    fun unselectTrait(t: Trait): Boolean {
        val removed = _creature.value!!.removeTrait(t.name)
        if (removed) _creature.value = _creature.value
        return removed
    }

    fun hasTrait(t: Trait): Boolean {
        return _creature.value!!.hasTrait(t.name)
    }
}